src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -203,28 +204,62 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
                 return;
             }
 
+            // When clients send acks we know for sure they received and applied the batch.
+            // We send batches right away, and queue them until we receive an ACK.
+            // If one or more client ACKs get lost we might receive an ack for a higher batch.
+            // We confirm all previous batches at that point (because receiving an ack is guarantee
+            // from the client that it has received and successfully applied all batches up to that point).
+
+            // If receive an ack for a previously rendered batch, its an error, as the messages are
+            // guranteed to be delivered in order, so a message for a render batch of 2 will never arrive
+            // after a message for a render batch for 3.
+            // If that were to be the case, it would just be enough to relax the checks here and simply skip
+            // the message.
+
+            // A batch might get lost when we send it to the client. If the client receives a newer batch when
+            // it was expecting a previous batch, it can simply force a reconnection.
+
             // Even though we're not on the renderer sync context here, it's safe to assume ordered execution of the following
             // line (i.e., matching the order in which we received batch completion messages) based on the fact that SignalR
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
-            if (!PendingRenderBatches.TryDequeue(out var entry) || entry.BatchId != incomingBatchId)
+            if (!PendingRenderBatches.TryPeek(out var entry) || incomingBatchId < entry.BatchId)
             {
                 HandleException(
                     new InvalidOperationException($"Received a notification for a rendered batch when not expecting it. Batch id '{incomingBatchId}'."));
```

Comment:
Is the following message more precise?

```suggestion
                    new InvalidOperationException($"Received a duplicate ACK for batch id '{incomingBatchId}'."));
```

If this statement does match the scenario, let's go for the more informative message.

And do you agree there's no possible *legitimate* scenario where a server receives a duplicate ACK? I'm struggling to think of a proof that the server can't receive a duplicate ACK, given that the client will ACK as many times as the server re-sends. For this to be robust, we're relying on guaranteed ordering of delivery not just within one SignalR connection but across successive connections established around reconnection.

If we're unable to reason precisely about this, let's consider just no-opping if we receive a duplicate ACK. We know it doesn't leave either side of the wire in a bad state.

Reply:
Also, should the exception type be whatever special internal exception type means "this can be triggered by bad external input so only log it as debug"?

Reply:
There are a bunch of places where we need to do that, so I'm not trying to tackle it as part of this PR. I want to go through the code-base and find all the places this happens.

If we decide to NOOP (which i'm ok with) then we don't need to use a special exception here.

Overall, this space is a "nightmare" to reason about.